Automattic/wp-calypso
# Checkout: Fix post-checkout behavior for 3DS cards

#### Changes proposed in this Pull Request

This fixes two problems related to post-checkout behavior when using 3DS cards:

**First, the thank-you page is seen before the order is complete.**

When using a 3DS card, first we get a reply from the transactions endpoint with the order ID and a request for auth. We send that request to Stripe directly to display a bank modal. When Stripe replies, we mark the transaction as complete, generate a thank-you URL, and visit it. The problem is that at this point the transaction is not quite complete; we still need Stripe to send a webhook to our servers to complete the order. The function that generates the thank-you URL is supposed to handle this by sending us to a "pending" page, but in order to do that it needs the order ID, and that was thrown away because we instead saved the response from Stripe.

This PR modifies the 3DS transaction process so that we instead save the response from the transactions endpoint.

**Second, the shopping-cart still shows the pre-checkout cart after we reach the final thank-you page.**

This happens because the only time that the cart is refreshed is just before we perform our post-checkout redirect; for 3DS cards, this goes to the "pending" page, but at that time the order is not yet complete and so the cart has not yet been emptied.

This PR modifies the pending page so that it reloads the cart when the order status changes.

Fixes 630-gh-Automattic/payments-shilling

#### Testing instructions

- Add a **new** plan to your cart and visit checkout. (This bug is not present for renewals.)
- Complete checkout with a 3DS card (if you have the API sandboxed you can use [Stripe test cards](https://stripe.com/docs/testing#three-ds-cards) like `4000000000003220`. It can either be a saved card or a new card.
- Confirm the purchase when the modal is displayed.
- Verify that you are redirected to a "pending" page for the order. (Note that on this page the masterbar shopping cart will be visible because it's still full; that's not ideal but it is a problem for another PR. I think that we may want to hide the entire masterbar on this page.)
- Verify that you are eventually redirected to a "Congratulations" page with the receipt ID in the URL.
- Verify that the shopping cart is empty (you should not see the masterbar shopping cart).
